include/nbl/builtin/hlsl/ieee754.hlsl

```diff
@@ -0,0 +1,128 @@
+	template<> int getMantissaBitCnt<uint32_t>() { return 23; }
+	template<> int getMantissaBitCnt<float64_t>() { return 52; }
+	template<> int getMantissaBitCnt<uint64_t>() { return 52; }
+
+	template <typename T>
+	int getExponentBias() { return 0xdeadbeefu; }
+	template<> int getExponentBias<float16_t>() { return 15; }
+	template<> int getExponentBias<uint16_t>() { return 15; }
+	template<> int getExponentBias<float>() { return 127; }
+	template<> int getExponentBias<uint32_t>() { return 127; }
+	template<> int getExponentBias<double>() { return 1023; }
+	template<> int getExponentBias<uint64_t>() { return 1023; }
+
+	template <typename T>
+	unsigned_integer_of_size<sizeof(T)>::type getExponentMask() { return 0xdeadbeefu; }
+	template<> unsigned_integer_of_size<2>::type getExponentMask<float16_t>() { return 0x7C00; }
+	template<> unsigned_integer_of_size<2>::type getExponentMask<uint16_t>() { return 0x7C00; }
+	template<> unsigned_integer_of_size<4>::type getExponentMask<float32_t>() { return 0x7F800000u; }
+	template<> unsigned_integer_of_size<4>::type getExponentMask<uint32_t>() { return 0x7F800000u; }
+	template<> unsigned_integer_of_size<8>::type getExponentMask<float64_t>() { return 0x7FF0000000000000ull; }
+	template<> unsigned_integer_of_size<8>::type getExponentMask<uint64_t>() { return 0x7FF0000000000000ull; }
+
+	template <typename T>
+	unsigned_integer_of_size<sizeof(T)>::type getMantissaMask() { return 0xdeadbeefu; }
+	template<> unsigned_integer_of_size<2>::type getMantissaMask<float16_t>() { return 0x03FF; }
+	template<> unsigned_integer_of_size<2>::type getMantissaMask<uint16_t>() { return 0x03FF; }
+	template<> unsigned_integer_of_size<4>::type getMantissaMask<float32_t>() { return 0x007FFFFFu; }
+	template<> unsigned_integer_of_size<4>::type getMantissaMask<uint32_t>() { return 0x007FFFFFu; }
+	template<> unsigned_integer_of_size<8>::type getMantissaMask<float64_t>() { return 0x000FFFFFFFFFFFFFull; }
+	template<> unsigned_integer_of_size<8>::type getMantissaMask<uint64_t>() { return 0x000FFFFFFFFFFFFFull; }
+
+	template <typename T>
+	uint32_t extractBiasedExponent(T x)
+	{
+		using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+		return bitfieldExtract<AsUint>(impl::castToUintType(x), getMantissaBitCnt<T>(), getExponentBitCnt<T>());
+	}
+
+	template <typename T>
+	int extractExponent(T x)
+	{
+		return int(extractBiasedExponent(x) - getExponentBias<T>());
+	}
+
+	template <typename T>
+	T replaceBiasedExponent(T x, uint32_t biasedExp)
+	{
+		return impl::replaceBiasedExponent(x, biasedExp);
+	}
+
+	// performs no overflow tests, returns x*exp2(n)
+	template <typename T>
+	T fastMulExp2(T x, int n)
+	{
+		return replaceBiasedExponent(x, extractBiasedExponent(x) + uint32_t(n));
+	}
+
+	template <typename T>
+	unsigned_integer_of_size<sizeof(T)>::type extractMantissa(T x)
+	{
+		using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+		return (impl::castToUintType(x) & getMantissaMask<T>());
+	}
```

Comment:
maybe these could be member methods of `ieee754::traits`

Reply:
why? no need imo